File: killer_sudoku/solver/engine/solver_engine.py
"""SolverEngine — main loop, apply_eliminations, trigger routing.

The engine builds a trigger → [rule] map at startup. apply_eliminations
routes BoardEvents to the work queue. The main loop pops items, skips
stale unit-scoped items via version tracking, calls rule.apply(), and
feeds returned Eliminations back through apply_eliminations.

GLOBAL rules (e.g. X-Wing) are processed through the same loop — there
is no separate run_global_rules function. When the queue contains only
GLOBAL items they are dequeued and applied normally; their eliminations
feed back into apply_eliminations to trigger further unit-scoped rules.
"""

from __future__ import annotations

import time
from typing import TYPE_CHECKING

from killer_sudoku.api.schemas import AutoMutation
from killer_sudoku.solver.engine.board_state import BoardState
from killer_sudoku.solver.engine.hint import HintResult
from killer_sudoku.solver.engine.rule import RuleContext, RuleStats, SolverRule
from killer_sudoku.solver.engine.types import Cell, Elimination, Trigger, UnitKind
from killer_sudoku.solver.engine.work_queue import SolverQueue
from killer_sudoku.solver.equation import sol_sums

if TYPE_CHECKING:
    from killer_sudoku.solver.engine.types import BoardEvent


def _filter_sum_range(
    cells: list[Cell],
    total: int,
    candidates: list[list[set[int]]],
) -> list[Elimination]:
    """Return eliminations for a non-burb sum constraint using range filtering.

    For each cell and each candidate digit, checks whether the digit can
    participate in an assignment summing to total given the current candidate
    ranges of the other cells.  Does NOT assume digit distinctness — safe for
    non-burb cell sets where cells may legally share a digit.
    """
    elims: list[Elimination] = []
    for i, (r, c) in enumerate(cells):
        others = [cells[j] for j in range(len(cells)) if j != i]
        min_others = sum(
            min(candidates[or_][oc]) for or_, oc in others if candidates[or_][oc]
        )
        max_others = sum(
            max(candidates[or_][oc]) for or_, oc in others if candidates[or_][oc]
        )
        for d in list(candidates[r][c]):
            if not (min_others + d <= total <= max_others + d):
                elims.append(Elimination(cell=(r, c), digit=d))
    return elims


def _filter_sum_constraint(
    cells: list[Cell],
    total: int,
    candidates: list[list[set[int]]],
) -> list[Elimination]:
    """Return eliminations for a newly-derived burb sum constraint.

    Computes all k-digit combinations summing to total (via sol_sums), maps
    each against current candidates using backtracking, and eliminates any
    (cell, digit) pair absent from every feasible assignment.

    Only call this for burb cell sets (single row/col/box), which guarantee
    digit distinctness — the required precondition for sol_sums.
    """
    solns = sol_sums(len(cells), 0, total)
    if not solns:
        return []

    # Sort most-constrained first for backtracking efficiency
    cand_union = frozenset().union(*solns)
    sorted_cells = sorted(
        cells, key=lambda rc: len(candidates[rc[0]][rc[1]] & cand_union)
    )

    per_cell_possible: dict[Cell, set[int]] = {c: set() for c in cells}

    def _bt(idx: int, remaining: set[int]) -> bool:
        if idx == len(sorted_cells):
            return True
        r, c = sorted_cells[idx]
        found = False
        for d in list(candidates[r][c] & remaining):
            remaining.discard(d)
            if _bt(idx + 1, remaining):
                per_cell_possible[(r, c)].add(d)
                found = True
            remaining.add(d)
        return found

    for soln in solns:
        _bt(0, set(soln))

    elims: list[Elimination] = []
    for cell in cells:
        r, c = cell
        for d in list(candidates[r][c]):
            if d not in per_cell_possible.get(cell, set()):
                elims.append(Elimination(cell=cell, digit=d))
    return elims


def _unit_kind(unit_id: int) -> UnitKind:
    """Determine the UnitKind from a global unit ID."""
    if unit_id < 9:
        return UnitKind.ROW
    if unit_id < 18:
        return UnitKind.COL
    if unit_id < 27:
        return UnitKind.BOX
    return UnitKind.CAGE


def _default_as_hints(
    rule: SolverRule, eliminations: list[Elimination]
) -> list[HintResult]:
    """Fallback hint for rules without meaningful as_hints() implementations."""
    if not eliminations:
        return []
    return [
        HintResult(
            rule_name=rule.name,
            display_name=rule.name,
            explanation=f"{rule.name} eliminated {len(eliminations)} candidate(s).",
            highlight_cells=frozenset(e.cell for e in eliminations),
            eliminations=eliminations,
        )
    ]


def _dedup_hints(hints: list[HintResult]) -> list[HintResult]:
    """Remove hints whose eliminations are entirely covered by earlier hints.

    Placement hints (placement is not None) are always kept.
    Preserves original order; first hint wins for any (cell, digit) pair.
    """
    seen_elims: set[tuple[Cell, int]] = set()
    seen_placements: set[tuple[int, int, int]] = set()
    result: list[HintResult] = []
    for h in hints:
        if h.placement is not None:
            if h.placement not in seen_placements:
                seen_placements.add(h.placement)
                result.append(h)
            continue
        new_elims = [e for e in h.eliminations if (e.cell, e.digit) not in seen_elims]
        if new_elims:
            for e in new_elims:
                seen_elims.add((e.cell, e.digit))
            result.append(
                HintResult(
                    rule_name=h.rule_name,
                    display_name=h.display_name,
                    explanation=h.explanation,
                    highlight_cells=h.highlight_cells,
                    eliminations=new_elims,
                    placement=h.placement,
                )
            )
    return result


class SolverEngine:
    """Pull-with-dirty-tracking propagation engine.

    Constructs a trigger → [rule] map at startup. apply_eliminations routes
    BoardEvents to matching rules in the priority queue. The main loop pops
    work items, skips stale unit-scoped items, calls rule.apply(), and feeds
    eliminations back through apply_eliminations.
    """

    def __init__(
        self,
        board: BoardState,
        rules: list[SolverRule],
        *,
        linear_system_active: bool = True,
        hint_rules: frozenset[str] = frozenset(),
    ) -> None:
        self.board = board
        self.queue: SolverQueue = SolverQueue()
        self.stats: dict[str, RuleStats] = {r.name: RuleStats() for r in rules}
        self._trigger_map: dict[Trigger, list[SolverRule]] = {t: [] for t in Trigger}
        self._linear_system_active = linear_system_active
        self._hint_rules = hint_rules
        self.pending_hints: list[HintResult] = []
        self.applied_mutations: list[AutoMutation] = []
        for rule in rules:
            for trigger in rule.triggers:
                self._trigger_map[trigger].append(rule)

    def apply_eliminations(self, eliminations: list[Elimination]) -> None:
        """Apply eliminations to board and route resulting events to the queue.

        For CELL_DETERMINED events, also calls LinearSystem.substitute_cell
        to propagate delta-pair constraints and recursively applies any new
        eliminations produced.
        """
        for elim in eliminations:
            r, c = elim.cell
            if elim.digit not in self.board.candidates[r][c]:
                continue
            if len(self.board.candidates[r][c]) <= 1:
                continue  # would raise NoSolnError; digit not in cands anyway
            events = self.board.remove_candidate(r, c, elim.digit)
            self._route_events(events, r, c)

    def _route_events(self, events: list[BoardEvent], src_r: int, src_c: int) -> None:
        """Route BoardEvents to the work queue; handle LinearSystem substitution.

        GLOBAL rules are re-scheduled on every board change so they get another
        pass after lower-priority rules have reduced the candidate set further.
        The deduplication in enqueue_global ensures at most one pending entry
        per rule at any time.
        """
        for event in events:
            if event.trigger == Trigger.CELL_DETERMINED:
                cell = event.payload
                assert isinstance(cell, tuple)
                val = event.hint_digit
                assert val is not None
                if self._linear_system_active:
                    # Propagate through LinearSystem delta pairs
                    new_elims = self.board.linear_system.substitute_cell(cell, val)
                    if new_elims:
                        self.apply_eliminations(new_elims)
                    # Dynamic RREF: reduce live rows and emit new sum constraints.
                    # This replicates the old solver's reduce_equns feedback loop:
                    # each determined cell can tighten multi-cell sum equations,
                    # producing new virtual cage constraints not derivable at startup.
                    new_constraints = self.board.linear_system.substitute_live_rows(
                        cell, val
                    )
                    for vcells, vtotal, distinct in new_constraints:
                        cell_list = list(vcells)
                        if len(cell_list) == 1:
                            # Single-cell constraint: directly determine the cell
                            lone = cell_list[0]
                            lr, lc = lone
                            for d in range(1, 10):
                                if d != vtotal and d in self.board.candidates[lr][lc]:
                                    self.apply_eliminations(
                                        [Elimination(cell=lone, digit=d)]
                                    )
                        elif distinct:
                            # Burb sum constraint: full sol_sums backtracking filter
                            sum_elims = _filter_sum_constraint(
                                cell_list, vtotal, self.board.candidates
                            )
                            if sum_elims:
                                self.apply_eliminations(sum_elims)
                        else:
                            # Non-burb sum constraint: range-only filter
                            sum_elims = _filter_sum_range(
                                cell_list, vtotal, self.board.candidates
                            )
                            if sum_elims:
                                self.apply_eliminations(sum_elims)
                # Route to CELL_DETERMINED rules (recognition, e.g. NakedSingle)
                for rule in self._trigger_map[Trigger.CELL_DETERMINED]:
                    self.queue.enqueue_cell(0, rule, cell, Trigger.CELL_DETERMINED, val)
                # Route to CELL_SOLVED rules (peer cleanup)
                for rule in self._trigger_map[Trigger.CELL_SOLVED]:
                    self.queue.enqueue_cell(0, rule, cell, Trigger.CELL_SOLVED, val)
            elif event.trigger == Trigger.SOLUTION_PRUNED:
                uid = event.payload
                assert isinstance(uid, int)
                for rule in self._trigger_map[Trigger.SOLUTION_PRUNED]:
                    self.queue.enqueue_unit(
                        rule.priority,
                        rule,
                        uid,
                        self.board.unit_versions[uid] - 1,
                        Trigger.SOLUTION_PRUNED,
                        None,
                    )
            else:
                uid = event.payload
                assert isinstance(uid, int)
                kind = _unit_kind(uid)
                for rule in self._trigger_map[event.trigger]:
                    if not rule.unit_kinds or kind in rule.unit_kinds:
                        self.queue.enqueue_unit(
                            rule.priority,
                            rule,
                            uid,
                            self.board.unit_versions[uid] - 1,
                            event.trigger,
                            event.hint_digit,
                        )

            # Re-schedule all GLOBAL rules whenever the board changes so they
            # can exploit new patterns created by lower-priority rule cascades.
            for rule in self._trigger_map[Trigger.GLOBAL]:
                self.queue.enqueue_global(rule.priority, rule)

    def _seed_initial_state(self) -> None:
        """Enqueue rules for all units to process the initial puzzle state.

        The engine is event-driven: rules only fire in response to candidate
        removals. Without this seed, puzzles with no LinearSystem initial
        eliminations would never trigger any rule. Seeding with unit_version=-1
        ensures no item is ever considered stale (unit_versions start at 0).

        Only COUNT_DECREASED and SOLUTION_PRUNED triggers are seeded. Triggers
        that require a specific hint_digit (COUNT_HIT_ONE, COUNT_HIT_TWO) and
        cell-level triggers (CELL_DETERMINED, GLOBAL) are not seeded here — they
        fire naturally as eliminations propagate from the seeded rules.
        """
        seed_triggers = frozenset({Trigger.COUNT_DECREASED, Trigger.SOLUTION_PRUNED})
        for unit in self.board.units:
            for trigger in seed_triggers:
                for rule in self._trigger_map[trigger]:
                    if not rule.unit_kinds or unit.kind in rule.unit_kinds:
                        self.queue.enqueue_unit(
                            rule.priority,
                            rule,
                            unit.unit_id,
                            -1,
                            trigger,
                            None,
                        )

    def solve(self) -> BoardState:
        """Run the main loop until no progress remains. Return the board state."""
        self.applied_mutations = []
        self.pending_hints = []
        # Seed all rules for all matching units (initial state propagation)
        self._seed_initial_state()
        # Enqueue GLOBAL sentinel for initial pass
        for rule in self._trigger_map[Trigger.GLOBAL]:
            self.queue.enqueue_global(rule.priority, rule)

        while not self.queue.empty():
            item = self.queue.pop()
            # CELL_DETERMINED and GLOBAL items are never version-skipped
            if item.is_stale(self.board.unit_versions):
                continue

            unit = self.board.units[item.unit_id] if item.unit_id is not None else None
            ctx = RuleContext(
                unit=unit,
                cell=item.cell,
                board=self.board,
                hint=item.trigger,
                hint_digit=item.hint_digit,
            )
            t0 = time.monotonic_ns()
            eliminations = item.rule.apply(ctx)
            elapsed = time.monotonic_ns() - t0
            self.stats[item.rule.name].record(eliminations, elapsed)

            if item.rule.name in self._hint_rules:
                hints = item.rule.as_hints(ctx, eliminations)
                self.pending_hints.extend(hints)
            elif eliminations:
                for e in eliminations:
                    self.applied_mutations.append(
                        AutoMutation(
                            rule_name=item.rule.name,
                            type="candidate_removed",
                            row=e.cell[0],
                            col=e.cell[1],
                            digit=e.digit,
                        )
                    )
                self.apply_eliminations(eliminations)

        self.pending_hints = _dedup_hints(self.pending_hints)
        return self.board

File: web/src/engine/solverEngine.ts
/**
 * SolverEngine — main loop, apply_eliminations, trigger routing.
 *
 * Mirrors Python's `killer_sudoku.solver.engine.solver_engine` module.
 *
 * Pull-with-dirty-tracking propagation engine. Constructs a trigger → [rule]
 * map at startup. apply_eliminations routes BoardEvents to the work queue.
 * The main loop pops items, skips stale unit-scoped items, calls rule.apply(),
 * and feeds eliminations back through apply_eliminations.
 *
 * Both classic and killer sudoku use the same engine. Classic puzzles simply
 * have empty cage_solns (cage_total = 0 → solSums returns []), so all
 * cage-specific rules become no-ops naturally.
 */

import { BoardState, CAGE_UNIT_OFFSET, KillerBoardState } from './boardState.js';
import type { HintResult } from './hint.js';
import type { RuleContext, RuleStats, SolverRule } from './rule.js';
import { makeRuleStats } from './rule.js';
import { NoSolnError } from '../solver/errors.js';
import { cellLabel } from './rules/_labels.js';
import {
  BoardEvent,
  Cell,
  cellKey,
  Elimination,
  Placement,
  SolutionElimination,
  Trigger,
  UnitKind,
  VirtualCageAddition,
  hasProgress,
} from './types.js';
import { isStale, SolverQueue } from './workQueue.js';

// ---------------------------------------------------------------------------
// Module-level helpers
// ---------------------------------------------------------------------------

export function toDisplayName(ruleName: string): string {
  return ruleName.replace(/([A-Z])/g, ' $1').trim();
}

function unitKindFromId(unitId: number): UnitKind {
  if (unitId < 9)  return UnitKind.ROW;
  if (unitId < 18) return UnitKind.COL;
  if (unitId < 27) return UnitKind.BOX;
  return UnitKind.CAGE;
}

function dedupHints(hints: HintResult[]): HintResult[] {
  const seenElims = new Set<string>();
  const seenPlacements = new Set<string>();
  const seenVc = new Set<string>();
  const result: HintResult[] = [];
  for (const h of hints) {
    if (h.virtualCageSuggestion !== null) {
      const key = JSON.stringify(h.virtualCageSuggestion);
      if (!seenVc.has(key)) { seenVc.add(key); result.push(h); }
      continue;
    }
    if (h.placement !== null) {
      const key = JSON.stringify(h.placement);
      if (!seenPlacements.has(key)) { seenPlacements.add(key); result.push(h); }
      continue;
    }
    const newElims = h.eliminations.filter(e => {
      const k = `${e.cell[0]},${e.cell[1]}:${e.digit}`;
      if (seenElims.has(k)) return false;
      seenElims.add(k);
      return true;
    });
    if (newElims.length > 0)
      result.push({ ...h, eliminations: newElims });
  }
  return result;
}

// ---------------------------------------------------------------------------
// SolverEngine
// ---------------------------------------------------------------------------

/** Options shared by SolverEngine and KillerSolverEngine. */
export interface SolverEngineOptions {
  hintRules?: ReadonlySet<string>;
  /** When provided, each rule application is checked: no rule may eliminate
   *  the correct solution digit from a cell where it is still a candidate.
   *  When `onViolation` is also provided, violations call it and suppress the
   *  rule result instead of throwing. When `onViolation` is null, violations
   *  throw NoSolnError (backward-compatible). */
  goldenSolution?: readonly (readonly number[])[] | null;
  /** Called when a rule produces an elimination that contradicts the golden
   *  solution. The engine suppresses the entire rule result (no board mutation)
   *  and continues. Only has effect when `goldenSolution` is also set. */
  onViolation?: ((ruleName: string, offending: readonly Elimination[]) => void) | null;
}

export class SolverEngine {
  readonly board: BoardState;
  readonly queue: SolverQueue;
  readonly stats: Map<string, RuleStats>;

  pendingHints: HintResult[] = [];
  appliedMutations: Array<{ruleName: string; type: string; [k: string]: unknown}> = [];
  appliedPlacements: Placement[] = [];
  appliedVirtualCages: VirtualCageAddition[] = [];

  protected readonly _triggerMap: Map<Trigger, SolverRule[]>;
  protected readonly _ruleIndex: Map<SolverRule, number>;
  private readonly _hintRules: ReadonlySet<string>;
  protected readonly _goldenSolution: readonly (readonly number[])[] | null;
  protected readonly _onViolation: ((ruleName: string, offending: readonly Elimination[]) => void) | null;
  /** True once a violation has been reported in the current solve() pass. Only the
   *  first violating rule is reported; subsequent violations are still suppressed
   *  but not reported (they may be cascades of the first bug). Reset at the start
   *  of each solve() call. */
  private _violationFired = false;

  constructor(
    board: BoardState,
    rules: SolverRule[],
    { hintRules = new Set<string>(), goldenSolution = null, onViolation = null }: SolverEngineOptions = {},
  ) {
    this.board = board;
    this.queue = new SolverQueue();
    this._ruleIndex = new Map(rules.map((r, i) => [r, i]));
    this.stats = new Map(rules.map(r => [r.name, makeRuleStats()]));
    this._hintRules = hintRules;
    this._goldenSolution = goldenSolution;
    this._onViolation = onViolation;

    this._triggerMap = new Map(Object.values(Trigger)
      .filter((v): v is Trigger => typeof v === 'number')
      .map(t => [t, [] as SolverRule[]]));
    for (const rule of rules)
      for (const trigger of rule.triggers)
        this._triggerMap.get(trigger)!.push(rule);
  }

  /** Linear-system propagation for a just-determined cell. No-op on a board
   *  with no LinearSystem; KillerSolverEngine overrides it to substitute the
   *  cell into the cage-sum equations and narrow live virtual-cage constraints. */
  protected _onCellDetermined(_cell: Cell, _val: number): void {}

  /** Reports (or throws, if no onViolation handler) when a forced/eliminated
   *  digit at (r, c) contradicts the golden solution. No-op if there is no
   *  golden solution or the digit matches it. */
  protected _checkAgainstGolden(ruleName: string, cell: Cell, digit: number): void {
    if (this._goldenSolution === null) return;
    const gold = this._goldenSolution[cell[0]]?.[cell[1]];
    if (gold === undefined || digit === gold) return;
    if (this._onViolation !== null) {
      this._onViolation(ruleName, [{ cell, digit: gold }]);
    } else {
      throw new NoSolnError(
        `${ruleName}: derived value ${digit} for r${cell[0] + 1}c${cell[1] + 1} contradicts golden solution ${gold}`,
      );
    }
  }

  applyEliminations(eliminations: readonly Elimination[]): void {
    for (const elim of eliminations) {
      const [r, c] = elim.cell;
      if (!this.board.cands(r, c).has(elim.digit)) continue;
      if (this.board.cands(r, c).size <= 1) continue;
      const events = this.board.removeCandidate(r, c, elim.digit);
      this._routeEvents(events, r, c);
    }
  }

  /**
   * Dispatch board events from a candidate removal to the work queue and linear system.
   *
   * CELL_DETERMINED events also propagate into the linear system (substitution + constraint
   * narrowing) and re-enqueue CELL_SOLVED listeners. SOLUTION_PRUNED and unit-scoped events
   * are routed to rules by trigger type and unit kind. Every event re-schedules all GLOBAL
   * rules so they see the latest board state.
   */
  private _routeEvents(events: BoardEvent[], _srcR: number, _srcC: number): void {
    for (const event of events) {
      if (event.trigger === Trigger.CELL_DETERMINED) {
        const cell = event.payload as Cell;
        const val = event.hintDigit!;
        this._onCellDetermined(cell, val);
        for (const rule of this._triggerMap.get(Trigger.CELL_DETERMINED) ?? [])
          this.queue.enqueueCell(0, rule, this._ruleIndex.get(rule)!, cell, Trigger.CELL_DETERMINED, val);
        for (const rule of this._triggerMap.get(Trigger.CELL_SOLVED) ?? [])
          this.queue.enqueueCell(0, rule, this._ruleIndex.get(rule)!, cell, Trigger.CELL_SOLVED, val);
      } else if (event.trigger === Trigger.SOLUTION_PRUNED) {
        const uid = event.payload as number;
        for (const rule of this._triggerMap.get(Trigger.SOLUTION_PRUNED) ?? [])
          this.queue.enqueueUnit(rule.priority, rule, this._ruleIndex.get(rule)!, uid,
            this.board.unitVersions[uid]! - 1, Trigger.SOLUTION_PRUNED, null);
      } else {
        const uid = event.payload as number;
        const kind = unitKindFromId(uid);
        for (const rule of this._triggerMap.get(event.trigger) ?? []) {
          if (rule.unitKinds.size === 0 || rule.unitKinds.has(kind))
            this.queue.enqueueUnit(rule.priority, rule, this._ruleIndex.get(rule)!, uid,
              this.board.unitVersions[uid]! - 1, event.trigger, event.hintDigit);
        }
      }
      // Re-schedule all GLOBAL rules on every board change
      for (const rule of this._triggerMap.get(Trigger.GLOBAL) ?? [])
        this.queue.enqueueGlobal(rule.priority, rule, this._ruleIndex.get(rule)!);
    }
  }

  private _seedInitialState(): void {
    const seedTriggers = new Set([Trigger.COUNT_DECREASED, Trigger.SOLUTION_PRUNED]);
    for (const unit of this.board.units) {
      for (const trigger of seedTriggers) {
        for (const rule of this._triggerMap.get(trigger) ?? []) {
          if (rule.unitKinds.size === 0 || rule.unitKinds.has(unit.kind))
            this.queue.enqueueUnit(rule.priority, rule, this._ruleIndex.get(rule)!, unit.unitId,
              -1, trigger, null);
        }
      }
    }
  }

  /** Cage-solution pruning for a rule's `solutionEliminations`. No-op on a board
   *  with no cage solutions; KillerSolverEngine overrides it to splice the pruned
   *  solution out of `cageSolns`, bump the cage's unit version, and enqueue
   *  SOLUTION_PRUNED/GLOBAL follow-ups. */
  protected _onSolutionElimination(_se: SolutionElimination, _ruleName: string): void {}

  /**
   * Run the rule engine to a fixed point and return the (mutated) board.
   *
   * Drains the work queue, applying each rule in priority order. Rules in `_hintRules`
   * accumulate their results in `pendingHints` instead of mutating the board. All other
   * rules write eliminations, placements, solution eliminations, and virtual cage
   * additions directly to the board and to `appliedMutations`/`appliedPlacements`/
   * `appliedVirtualCages`. Resets all accumulators on each call. Deduplicates
   * `pendingHints` before returning.
   */
  solve(): BoardState {
    this.appliedMutations = [];
    this.appliedPlacements = [];
    this.appliedVirtualCages = [];
    this.pendingHints = [];
    this._violationFired = false;

    this._seedInitialState();
    for (const rule of this._triggerMap.get(Trigger.GLOBAL) ?? [])
      this.queue.enqueueGlobal(rule.priority, rule, this._ruleIndex.get(rule)!);

    while (!this.queue.empty()) {
      const item = this.queue.pop();
      if (isStale(item, this.board.unitVersions)) continue;

      const unit = item.unitId !== null ? this.board.units[item.unitId] ?? null : null;
      const ctx: RuleContext = {
        unit,
        cell: item.cell,
        board: this.board,
        hint: item.trigger,
        hintDigit: item.hintDigit,
      };

      const t0 = performance.now();
      const result = item.rule.apply(ctx);
      const elapsed = (performance.now() - t0) * 1e6; // to nanoseconds
      const stats = this.stats.get(item.rule.name)!;
      stats.calls++;
      if (hasProgress(result)) stats.progress++;
      stats.eliminations += result.eliminations.length;
      stats.elapsedNs += elapsed;

      if (this._hintRules.has(item.rule.name)) {
        // Golden check for hint rules: a hint whose eliminations contradict the
        // golden solution is unsound. Suppress it and report the violation so the
        // rule is disabled for the session.
        // Unlike always-apply rules, hint rules cannot cascade eliminations into
        // other rules, so _violationFired is not set here — every bad hint rule
        // is reported independently.
        if (this._goldenSolution !== null && result.eliminations.length > 0) {
          const offending = result.eliminations.filter(e => {
            const [r, c] = e.cell;
            const gold = this._goldenSolution![r]?.[c];
            return gold !== undefined && e.digit === gold && this.board.cands(r, c).has(gold);
          });
          if (offending.length > 0) {
            if (this._onViolation !== null) this._onViolation(item.rule.name, offending);
            continue; // suppress the bad hint
          }
        }
        this.pendingHints.push(...item.rule.asHints(ctx, result.eliminations));
      } else {
        if (this._goldenSolution !== null && result.virtualCageAdditions.length > 0) {
          const offending = result.virtualCageAdditions.find(vca => {
            const goldSum = vca.cells.reduce((sum, [r, c]) => sum + this._goldenSolution![r]![c]!, 0);
            return goldSum !== vca.total;
          });
          if (offending !== undefined) {
            if (this._onViolation !== null) {
              // Report only the first violation per solve() pass — subsequent
              // violations may be cascades of the first bug.
              if (!this._violationFired) {
                this._onViolation(item.rule.name, []);
                this._violationFired = true;
              }
              // Always suppress the entire rule result regardless of whether
              // the violation was reported.
              continue;
            } else {
              const goldSum = offending.cells.reduce((sum, [r, c]) => sum + this._goldenSolution![r]![c]!, 0);
              throw new NoSolnError(
                `${item.rule.name}: virtual cage ${offending.cells.map(c => cellLabel(c)).join('+')} = ${offending.total} ` +
                `contradicts golden solution (sums to ${goldSum})`,
              );
            }
          }
        }

        if (result.eliminations.length > 0) {
          if (this._goldenSolution !== null) {
            const offending = result.eliminations.filter(e => {
              const [r, c] = e.cell;
              const gold = this._goldenSolution![r]?.[c];
              return gold !== undefined && e.digit === gold && this.board.cands(r, c).has(gold);
            });
            if (offending.length > 0) {
              if (this._onViolation !== null) {
                // Report only the first violation per solve() pass — subsequent
                // violations may be cascades of the first bug.
                if (!this._violationFired) {
                  this._onViolation(item.rule.name, offending);
                  this._violationFired = true;
                }
                // Always suppress the entire rule result regardless of whether
                // the violation was reported.
                continue;
              } else {
                const [r, c] = offending[0]!.cell;
                const gold = offending[0]!.digit;
                throw new NoSolnError(
                  `${item.rule.name}: would eliminate correct digit ${gold} from r${r + 1}c${c + 1}`,
                );
              }
            }
          }
          for (const e of result.eliminations)
            this.appliedMutations.push({ ruleName: item.rule.name, type: 'candidate_removed',
              row: e.cell[0], col: e.cell[1], digit: e.digit });
          this.applyEliminations(result.eliminations);
        }
        for (const p of result.placements) {
          this.appliedPlacements.push(p);
          this.appliedMutations.push({ ruleName: item.rule.name, type: 'placement',
            row: p.cell[0], col: p.cell[1], digit: p.digit });
        }
        for (const se of result.solutionEliminations)
          this._onSolutionElimination(se, item.rule.name);
        for (const vca of result.virtualCageAdditions) {
          if (!(this.board instanceof KillerBoardState)) continue;
          this.board.addVirtualCage(vca.cells, vca.total, []);
          this.board.linearSystem.pendingVirtualCages.shift();

          // Seed COUNT_DECREASED/SOLUTION_PRUNED for the new unit, mirroring
          // _seedInitialState, so cage rules evaluate it within this pass.
          const newUnitId = this.board.units.length - 1;
          for (const trigger of [Trigger.COUNT_DECREASED, Trigger.SOLUTION_PRUNED]) {
            for (const rule of this._triggerMap.get(trigger) ?? []) {
              if (rule.unitKinds.size === 0 || rule.unitKinds.has(UnitKind.CAGE))
                this.queue.enqueueUnit(rule.priority, rule, this._ruleIndex.get(rule)!,
                  newUnitId, -1, trigger, null);
            }
          }

          this.appliedVirtualCages.push(vca);
          this.appliedMutations.push({ ruleName: item.rule.name, type: 'virtual_cage_added',
            cells: vca.cells, total: vca.total });
        }
      }
    }

    this.pendingHints = dedupHints(this.pendingHints);
    return this.board;
  }
}

// ---------------------------------------------------------------------------
// KillerSolverEngine — owns linear-system propagation and cage-solution pruning
// ---------------------------------------------------------------------------

export class KillerSolverEngine extends SolverEngine {
  override readonly board: KillerBoardState;

  constructor(board: KillerBoardState, rules: SolverRule[], opts: SolverEngineOptions = {}) {
    super(board, rules, opts);
    this.board = board;
  }

  protected override _onCellDetermined(cell: Cell, val: number): void {
    const newConstraints = this.board.linearSystem.substituteLiveRows(cell, val);
    if (newConstraints.length === 0) return;

    // Cell-set -> total for existing units, so a derived constraint is only
    // skipped as redundant when an existing unit covers the same cells *and*
    // sums to the same total. A matching cell-set with a different total is a
    // genuinely new constraint (e.g. it may contradict a corrupted cage total).
    const existingTotals = new Map<string, number>();
    for (const u of this.board.units) {
      const key = u.cells.map(cellKey).slice().sort().join('|');
      if (u.kind === UnitKind.CAGE) {
        const solns = this.board.cageSolns[u.unitId - CAGE_UNIT_OFFSET];
        if (solns && solns.length > 0) existingTotals.set(key, solns[0]!.reduce((a, b) => a + b, 0));
      } else {
        // Rows/cols/boxes always hold 1-9 exactly once, summing to 45.
        existingTotals.set(key, 45);
      }
    }
    for (const [vcells, vtotal, distinct] of newConstraints) {
      if (!distinct) continue;
      const cells = [...vcells] as Cell[];
      if (cells.length === 1) {
        this._checkAgainstGolden('DerivedVirtualCage', cells[0]!, vtotal);
      }
      const key = cells.map(cellKey).slice().sort().join('|');
      if (existingTotals.get(key) === vtotal) continue;
      existingTotals.set(key, vtotal);
      this.board.linearSystem.pendingVirtualCages.push({ cells, total: vtotal });
    }
  }

  protected override _onSolutionElimination(se: SolutionElimination, ruleName: string): void {
    const solns = this.board.cageSolns[se.cageIdx]!;
    const idx = solns.findIndex(s => s.length === se.solution.length && s.every((d, i) => d === se.solution[i]));
    if (idx < 0) return;
    solns.splice(idx, 1);
    const cageUnitId = CAGE_UNIT_OFFSET + se.cageIdx;
    this.board.unitVersions[cageUnitId]!++;
    this.appliedMutations.push({ ruleName, type: 'solution_eliminated', cageIdx: se.cageIdx, solution: se.solution });
    for (const rule of this._triggerMap.get(Trigger.SOLUTION_PRUNED) ?? [])
      this.queue.enqueueUnit(rule.priority, rule, this._ruleIndex.get(rule)!, cageUnitId,
        this.board.unitVersions[cageUnitId]! - 1, Trigger.SOLUTION_PRUNED, null);
    for (const rule of this._triggerMap.get(Trigger.GLOBAL) ?? [])
      this.queue.enqueueGlobal(rule.priority, rule, this._ruleIndex.get(rule)!);
  }
}
